config/config.go

```diff
@@ -0,0 +1,73 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/BurntSushi/toml"
+)
+
+type Config struct {
+	HomeDir string
+	BinDir  string
+	VarDir  string
+	RunDir  string
+
+	Telemetry struct {
+		Disable bool
+	}
+}
+
+type contextKey string
+
+var configContextKey = contextKey("config")
+
+func LoadDefault(cfg *Config) error {
+	if cfg.HomeDir == "" {
+		cfg.HomeDir = exoHome()
+	}
+
+	// TODO: make directory if necessary.
+	configFile := getExoPath(cfg.HomeDir, "config.toml", "EXO_CONFIG")
+	if err := loadFromFile(configFile, cfg); err != nil {
+		return err
+	}
+
+	if cfg.BinDir == "" {
+		cfg.BinDir = getExoPath(cfg.HomeDir, "bin", "EXO_BIN")
+	}
+	if cfg.RunDir == "" {
+		cfg.RunDir = getExoPath(cfg.HomeDir, "run", "EXO_RUN")
+	}
+	if cfg.VarDir == "" {
+		cfg.VarDir = getExoPath(cfg.HomeDir, "var", "EXO_VAR")
+	}
+
+	return nil
+}
+
+func MustLoadDefault(cfg *Config) {
+	if err := LoadDefault(cfg); err != nil {
+		panic(err)
+	}
+}
+
+func loadFromFile(filePath string, cfg *Config) error {
+	_, err := toml.DecodeFile(filePath, cfg)
+	if os.IsNotExist(err) {
+		if err = os.WriteFile(filePath, []byte{'\n'}, 0644); err != nil {
+			return fmt.Errorf("creating initial config file: %w", err)
+		}
+	}
+	return err
+}
+
+func WithConfig(ctx context.Context, cfg *Config) context.Context {
+	return context.WithValue(ctx, configContextKey, cfg)
+}
+
+func GetConfig(ctx context.Context) (*Config, bool) {
```

Comment:
you're not using this anywhere yet, do you expect to? If possible, would be nice to avoid context with/get value things as best we can and be explicit with dependency injection unless we need to cross interface boundaries we don't control.

Reply:
This started as part of the telemetry disabling. Explicit dependency injection would be nice (especially using something like wire), but I think that the context thing is good for now, then we can do a better job segregating things later. I mean, we should probably have a Telemetry struct and pass that in everywhere, but (especially with multiple entry points), it is a lot to wire up manually.

Reply:
Rethinking this. It would not be too much work to do a better job of this. I'll update the PR.